cdc/capture/capture.go

```diff
@@ -580,100 +582,17 @@ func (c *captureImpl) AsyncClose() {
 }
 
 // Drain removes tables in the current TiCDC instance.
-func (c *captureImpl) Drain(ctx context.Context) <-chan struct{} {
-	log.Info("draining capture, removing all tables on the capture",
-		zap.String("captureID", c.info.ID))
-
-	const drainInterval = 100 * time.Millisecond
-	done := make(chan struct{})
-	go func() {
-		defer close(done)
-		ticker := time.NewTicker(drainInterval)
-		defer ticker.Stop()
-		for {
-			complete := c.drainImpl(ctx)
-			if complete {
-				return
-			}
-			ticker.Reset(drainInterval)
-			select {
-			case <-ctx.Done():
-				// Give up when the context cancels. In the current
-				// implementation, it is caused TiCDC receives a second signal
-				// and begins force shutdown.
-				return
-			case <-ticker.C:
-			}
-		}
-	}()
-	return done
-}
-
-func (c *captureImpl) drainImpl(ctx context.Context) bool {
-	if !c.config.Debug.EnableSchedulerV3 {
-		// Skip drain as two phase scheduler is disabled.
-		return true
-	}
-
-	// Step 1, resign ownership.
-	o, _ := c.GetOwner()
-	if o != nil {
-		doneCh := make(chan error, 1)
-		query := &owner.Query{Tp: owner.QueryCaptures, Data: []*model.CaptureInfo{}}
-		o.Query(query, doneCh)
-		select {
-		case <-ctx.Done():
-		case err := <-doneCh:
-			if err != nil {
-				log.Warn("query capture count failed, retry", zap.Error(err))
-				return false
-			}
-		}
-		if len(query.Data.([]*model.CaptureInfo)) <= 1 {
-			// There is only one capture, the owner itself. It's impossible to
-			// resign owner nor move out tables, give up drain.
-			log.Warn("there is only one capture, skip drain")
-			return true
-		}
+func (c *captureImpl) Drain() {
```

Comment:
How about adding a for loop to make sure it resigns ownership?

Reply:
no need to wait for the ownership resigend, most of the time it's very fast.

When the user stop the cdc instance by sent signal, the normal `server.Close()` is guaranteed that the ownership resigned before exit.